opal/mca/mpool/memkind/mpool_memkind_component.c

```diff
@@ -229,37 +296,85 @@ static int mca_mpool_memkind_query (const char *hints, int *priority_out,
             }
-        } else if (0 == strcasecmp (key, "partition")) {
-            rc = mca_mpool_memkind_enum->value_from_string (mca_mpool_memkind_enum,
-                                                            value, &partition);
+        } else if (0 == strcasecmp (key, "policy")) {
+
+            rc = mca_mpool_memkind_policy_enum->value_from_string (mca_mpool_memkind_policy_enum,
+                                                                   value, (int *)&policy);
+            if (OPAL_SUCCESS != rc) {
+                opal_output_verbose (MCA_BASE_VERBOSE_WARN, mca_mpool_memkind_component.output,
+                                     "invalid memkind policy %s specified", value);
+            }
+
+        } else if (0 == strcasecmp (key, "type")) {
+
+            rc = mca_mpool_memkind_type_enum->value_from_string (mca_mpool_memkind_type_enum,
+                                                                 value, (int *)&type);
             if (OPAL_SUCCESS != rc) {
                 opal_output_verbose (MCA_BASE_VERBOSE_WARN, mca_mpool_memkind_component.output,
-                                     "invalid partition %s specified", value);
+                                     "invalid memkind type %s specified", value);
             }
 
-            partition_name = value;
+        } else if (0 == strcasecmp (key, "kind_bits")) {
+
+            rc = mca_mpool_memkind_kind_bits_enum->value_from_string (mca_mpool_memkind_kind_bits_enum,
+                                                                      value, (int *)&memkind_bits);
+            if (OPAL_SUCCESS != rc) {
+                opal_output_verbose (MCA_BASE_VERBOSE_WARN, mca_mpool_memkind_component.output,
+                                     "invalid memkind kind_bits %s specified", value);
+            }
         }
     }
 
-    if (-1 != partition) {
-        memkind_t kind;
+    /*
+     * now look for an existing module with matching policy, type, memkind bits
+     */
+
+    OPAL_LIST_FOREACH(item, &mca_mpool_memkind_component.module_list, 
+                      mca_mpool_memkind_module_le_t) {
+        if ((item->module.type == type) &&
+            (item->module.policy == policy) &&
+            (item->module.memkind_bits = memkind_bits)) {
+               found_module = &item->module.super;
+               break;
+        }
+    } 
 
-        my_priority = 0;
+    /*
+     * didn't find a matching module, try to create one 
+     */
 
-        if (!memkind_get_kind_by_partition (partition, &kind)) {
-            if (memkind_check_available (kind)) {
-                opal_output_verbose (MCA_BASE_VERBOSE_WARN, mca_mpool_memkind_component.output,
-                                     "kind %s not available", partition_name);
+    if (NULL == found_module) {
+        rc = memkind_create_kind(type, policy, memkind_bits, &kind);
+        if (MEMKIND_SUCCESS == rc) {
+
+            item = OBJ_NEW(mca_mpool_memkind_module_le_t);
+
```

Comment:
This class name is wrong. Shouldn't have _le in it unless you changed the struct name (looks like you didn't).

Reply:
Nevermind, I see it.

Reply:
But I don't see an OBJ instance.

Reply:
Or a constructor for that matter.